### ISSUE TITLE
Need your advice to train wheat head dataset with YOLOv4

### ISSUE THREAD

**Issue by Twinkle126** [AlexeyAB/darknet#6110]
If you have an issue with training - no-detections / Nan avg-loss / low accuracy:
    * what command do you use?
    * what dataset do you use?   
    * what Loss and mAP did you get?
    * show chart.png with Loss and mAP    
    * check your dataset - run training with flag `-show_imgs` i.e. `./darknet detector train ... -show_imgs` and look at the `aug_...jpg` images, do you see correct truth bounded boxes?
    * rename your cfg-file to txt-file and drag-n-drop (attach) to your message here
    * show content of generated files `bad.list` and `bad_label.list` if they exist
    * Read `How to train (to detect your custom objects)` and `How to improve object detection` in the Readme: https://github.com/AlexeyAB/darknet/blob/master/README.md
    * show such screenshot with info
```
./darknet detector test cfg/coco.data cfg/yolov4.cfg yolov4.weights data/dog.jpg
 CUDA-version: 10000 (10000), cuDNN: 7.4.2, CUDNN_HALF=1, GPU count: 1
 CUDNN_HALF=1
 OpenCV version: 4.2.0
 0 : compute_capability = 750, cudnn_half = 1, GPU: GeForce RTX 2070
net.optimized_memory = 0
mini_batch = 1, batch = 8, time_steps = 1, train = 0
   layer   filters  size/strd(dil)      input                output
```
Hi @AlexeyAB , 

I am working on a data containing images of wheat fields with bounding boxes for each identified wheat head.  Goal is to identify wheat heads. The bounding box detail is given in a csv with respect to each image and one sample image is as below-
![0a3cb453f](https://user-images.githubusercontent.com/36033503/86254715-d8527e80-bbd3-11ea-9727-45d6ab5495e6.jpg)

The training dataset has total 3422 images and I have prepared the dataset as you mentioned in the repo. In my second attempt I also triedbelow mentioned changes-
![small_object](https://user-images.githubusercontent.com/36033503/86254769-f4eeb680-bbd3-11ea-99c1-86dc608d1813.JPG)

But I am getting same training result in both cases- loss is not decreasing so this time I stopped the training in middle
![wheat_training_chart](https://user-images.githubusercontent.com/36033503/86254968-354e3480-bbd4-11ea-93ea-43964a1f4c7a.JPG)

My cfg file is as below-
[yolo-obj.txt](https://github.com/AlexeyAB/darknet/files/4858408/yolo-obj.txt)

Can you please look into this once and let me know my mistake?
In case you need any detail, let me know that also.

**Comment by AlexeyAB** [comment]
Loss is normal. Check the mAP.

**Comment by Twinkle126** [comment]
Ok, I will check with command `.\darknet.exe detector train data/obj.data cfg/yolo-obj.cfg yolo-obj_2000.weights -map` and let you know the result.

**Comment by Twinkle126** [comment]
@AlexeyAB , Till now chart is as below with map parameter-
![map_wheat](https://user-images.githubusercontent.com/36033503/86312627-c78c2200-bc40-11ea-896b-017ef4fdb4fe.JPG)

**Comment by Twinkle126** [comment]
Hi @AlexeyAB ,

Training is completed with map argument-
![map_done](https://user-images.githubusercontent.com/36033503/86371122-2a5fd680-bc9e-11ea-8185-23c5e67749a7.JPG)

And chart is as below-
![chart](https://user-images.githubusercontent.com/36033503/86371178-38155c00-bc9e-11ea-893c-807de41e441c.png)

Is it fine?
Can I do anything to improve the accuracy?
[yolo-obj.txt](https://github.com/AlexeyAB/darknet/files/4864925/yolo-obj.txt)

**Comment by AlexeyAB** [comment]
If resolution of your images higher than 608x608 then you can set width=704 height=704, then train and use it.

**Comment by Twinkle126** [comment]
@AlexeyAB Thanks!
I will try with width=704 height=704 and let you know the results.

**Comment by Twinkle126** [comment]
@AlexeyAB ,
Although image resolotion is 1024 x 1024 yet cannot use either 608 or 704 due to memory error-
![cuda_memory](https://user-images.githubusercontent.com/36033503/86387118-52583580-bcb0-11ea-9e59-ef894ed6e1f2.JPG)

subdivisions is already set to 64 in my cfg-file :(

**Comment by AlexeyAB** [comment]
change random=1 to resize=1.5 in cfg

**Comment by Twinkle126** [comment]
I have changed random=1 to resize=1.5 in cfg file and image width/height as 704.
With this change training is started :)
I will update you once it is completed.

**Comment by dsbyprateekg** [comment]
@AlexeyAB 
Training is complete now with resize=1.5 parameter.
It took around 36 hours to complete and result is as bellow-
![map_latest_wheat](https://user-images.githubusercontent.com/30830541/86537795-44ecb680-bf0f-11ea-8a41-d7c474bdc35f.JPG)

![chart](https://user-images.githubusercontent.com/30830541/86537808-5a61e080-bf0f-11ea-8834-efb970750a87.png)

Is the above result fine?

**Comment by AlexeyAB** [comment]
Yes.
Try to check the mAP with flag `./darknet detector map ... -iou_thresh 0.05` and show screenshot.

**Comment by dsbyprateekg** [comment]
@AlexeyAB 
I ran the command `.\darknet.exe detector map data/obj.data cfg/yolo-obj.cfg backup/wheat/yolo-obj_best.weights` and got below result-
![detect_map_wheat](https://user-images.githubusercontent.com/30830541/86555248-7ac08800-bf6d-11ea-98db-2436797d0d0e.JPG)

And with command `.\darknet.exe detector map data/obj.data cfg/yolo-obj.cfg backup/wheat/yolo-obj_best.weights -iou_thresh 0.05` got below result-
![detect_map_wheat_thres](https://user-images.githubusercontent.com/30830541/86556173-27036e00-bf70-11ea-9fec-03a539fbffe0.JPG)

**Comment by AlexeyAB** [comment]
AP50 is increased from 90% to 93%
AP05 is increased from 93% to 97.5% (the error dropped from 7% to 2.5%, about 3x times) - **this is very good result.**

----

If you want to increase accuracy more, just try to train with width=1024 height=1024 in cfg on GPU with more GPU-RAM.

**Comment by dsbyprateekg** [comment]
Thanks for the confirmation @AlexeyAB 
Since I have limited RAM, I will continue my work with this :)

**Comment by NitinDatta8** [comment]
Hello guys,
Even I am trying to detect wheat heads but my bounding boxes are projected wrong on the image
Below I have pasted the screenshot of one of the coordinates
It is [class, x-center, y-center, w,h]
Can someone please help me
Thank you
![Screenshot](https://user-images.githubusercontent.com/36504819/87071515-f6edf080-c237-11ea-8d85-98e8905d3b04.png)

**Comment by stephanecharette** [comment]
> Can someone please help me

Help you to do what?  You pasted a screenshot of some coordinates.  Not much we can do with your screenshot.

**Comment by NitinDatta8** [comment]
Hi,
I'm so sorry l will update it ASAP.
Regards
Nitin

**Comment by Hwijune** [comment]
Have you checked anchor box and ground truth?

**Comment by NitinDatta8** [comment]
[UPDATE]
Hi guys, 
I guess I found my mistake...
the class should be in integer  and not float I changed the label files and now I tested it is making correct bounding boxes
